README.md
# demo

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2022 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2022-12-20
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2022-12-20
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/demo/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/demo/compare/0.1.0...0.1.1


src/demo/test.clj
(ns demo.test)
(defn foo1
  "I don't do a whole lot."
  [x]
  (println x "Hello, World!"))

(println (foo1 "Hello Clojure \n"))


src/demo/futures.clj
(ns demo.futures
  (:use clojure.pprint))
(require '[clojure.repl :refer [demunge]])
;并发操作
;    协调  不协调
;同步 refs atoms
;异步 /    agents

(defmacro futures
  [n & exprs]
  (vec (for [_ (range n)
             expr exprs]
         `(future ~expr))))

(defmacro wait-futures
  [& args]
  `(doseq [f# (futures ~@args)]
     @f#))

;Atoms;;;;;;;;;;;;;

;先比较再设值（协调。阻塞）
;没有其它线程会看到atom的中间结果。
;swap! 会重试！
(def sarah (atom {:name "Sarah" :age 25 :wears-glasses? false}))
(swap! sarah update-in [:age] + 3)
(swap! sarah (comp #(update-in % [:age] inc)
                   #(assoc % :wears-glasses? true)))

;trying 5 会重试！
(def xs (atom #{1 2 3}))
(wait-futures 1
              (swap! xs (fn [v]
                            (Thread/sleep 250)
                            (println "trying 4")
                            (conj v 4)))
              (swap! xs (fn [v]
                          (Thread/sleep 500)
                          (println "trying 5")
                          (conj v 5))))

(def x (atom 2000))
;= #'user/x
(wait-futures 1
              (swap! x #(Thread/sleep %))
              (println @x)
              )

;不管当前值是什么，重新设值
(reset! x :y)

;Watches
(defn echo-watch
  [key identity old new]
  (println key old "=>" new))

(def sarah (atom {:name "Sarah" :age 25}))

(add-watch sarah :echo echo-watch)

(swap! sarah update-in [:age] inc)

(add-watch sarah :echo2 echo-watch)

(swap! sarah update-in [:age] inc)

(remove-watch sarah :echo2)

(swap! sarah update-in [:age] inc)

(reset! sarah @sarah)

;log watchs
(def history (atom ()))

(defn log->list
  [dest-atom key source old new]
  (when (not= old new)
    (swap! dest-atom conj new)))

(def sarah (atom {:name "Sarah" :age 25}))

(add-watch sarah :recode (partial log->list history))

(swap! sarah update-in [:age] inc)
(swap! sarah update-in [:age] inc)

(swap! sarah identity)

(swap! sarah assoc :wears-glasses? true)
(swap! sarah update-in [:age] inc)

(pprint @history)

;(defn log->db
;  [db-id identity old new]
;  (when (not= old new)
;    (let [db-connection (get-connection db-id)]
;      ...)))

;Validators
(def n (atom 1 :validator pos?))
(swap! n + 500)
(swap! n - 1000)

(def sarah (atom {:name "Sarah" :age 25}))
(set-validator! sarah :age)
(swap! sarah dissoc :age)

(set-validator! sarah #(or (:age %)
                           (throw (IllegalStateException. "People must have `:age`s!"))))

(swap! sarah dissoc :age)

;Refs;;;;;;;;;;;;;;;

(defn character
  [name & {:as opts}]
   (ref (merge {:name name :items #{} :health 500}
               opts)))

(def smaug (character "Smaug" :health 500 :strength 400 :items (set (range 50))))
(def bilbo (character "Bilbo" :health 100 :strength 100))
(def gandalf (character "Gandalf" :health 75 :mana 750))
;bilbo,gandalf 抢夺smaug :item所有武器！
;dosync 规定 STM 边界
;alter,commute,ref-set 对 ref 进行修改
;loot函数负责抢夺武器。
(defn loot
  [from to]
  (dosync
    (when-let [item (first (:items @from))]
      (alter to update-in [:items] conj item)
      (alter from update-in [:items] disj item))))

(wait-futures 1
              (while (loot smaug bilbo))
              (while (loot smaug gandalf)))

;验证正确性
(map (comp count :items deref) [bilbo gandalf] )
(filter (:items @bilbo) (:items @gandalf))

;COMMUTE 改善性能。commute提交时用最新值。不进行初始值与新值校验。
;commute 的函数符合交换率，就是可以没有前后之分。如+。先加后加谁都一样。
(def x (ref 0))
;232"
(time (wait-futures 5
                    (dotimes [_ 1000]
                      (dosync (alter x + (apply + (range 1000)))))
                    (dotimes [_ 1000]
                      (dosync (alter x + (apply + (range 1000)))))))
;59"
(time (wait-futures 5
                    (dotimes [_ 1000]
                      (dosync (commute x + (apply + (range 1000)))))
                    (dotimes [_ 1000]
                      (dosync (commute x + (apply + (range 1000)))))))
;ref加武器，没有顺序；取武器有顺序
(defn flawed-loot
  [from to]
  (dosync
    (when-let [item (first (:items @from))]
      (commute to update-in [:items] conj item)
      (alter from update-in [:items] disj item))))

(def smaug (character "Smaug" :health 500 :strength 400 :items (set (range 50))))
(def bilbo (character "Bilbo" :health 100 :strength 100))
(def gandalf (character "Gandalf" :health 75 :mana 750))

(wait-futures 1
              (while (flawed-loot smaug bilbo))
              (while (flawed-loot smaug gandalf)))

;验证正确性
(map (comp count :items deref) [bilbo gandalf] )
(filter (:items @bilbo) (:items @gandalf))

;;other actions
(defn attack
  [aggressor target]
  (dosync
    (let [damage (* (rand 0.1) (:strength @aggressor))]
      (commute target update-in [:health] #(max 0 (- % damage))))))

(defn heal
  [healer target]
  (dosync
    (let [aid (* (rand 0.1) (:mana @healer))]
      (when (pos? aid)
        (commute healer update-in [:mana] - (max 5 (/ aid 5)))
        (commute target update-in [:health] + aid)))))

;;demo
(def alive? (comp pos? :health))

(defn play
  [character action other]
  (while (and (alive? @character)
              (alive? @other)
              (action character other))
    (Thread/sleep (rand-int 50))))

;;action
(wait-futures 1
              (play bilbo attack smaug)
              (play smaug attack bilbo))

(map (comp :health deref) [smaug bilbo])

;; 多人战斗
(dosync
  (alter smaug assoc :health 500)
  (alter bilbo assoc :health 100)
  (alter gandalf assoc :mana 750))

(wait-futures 1
              (play bilbo attack smaug)
              (play smaug attack bilbo)
              (play gandalf heal bilbo))

(map (comp #(select-keys % [:name :health :mana]) deref) [smaug bilbo gandalf])

;;修改ref. 与alter 相同 都会重试！
(dosync (ref-set bilbo {:name "Bilbo"}))
(dosync (alter bilbo (constantly {:name "Bilbo"})))

;;校验器
(defn- enforce-max-health
  [{:keys [name health]}]
  (fn [character-data]
    (or (<= (:health character-data) health)
        (throw (IllegalStateException. (str name " is already st max health!"))))))

(defn character
  [name & {:as opts}]
  (let [cdata (merge {:name name :items #{} :health 500}
                     opts)
        cdata (assoc cdata :max-health (:health cdata))
        validators (list* (enforce-max-health cdata)
                          (:validators cdata))]
    (ref (dissoc cdata :validators)
         :validator #(every? (fn [v] (v %)) validators))))

(def bilbo (character "Bilbo" :health 100 :strength 100))

(heal gandalf bilbo)

(get-validator bilbo)
(-> (get-validator bilbo)  ; get-validator returns the function object
    class                ; get the Class instance
    .getSimpleName       ; access the inner class name, no package
    demunge              ; demunge is the opposite of munge
    symbol)

;;改善heal
(defn heal
  [healer target]
  (dosync
    (let [aid (min (* (rand 0.1) (:mana @healer))
                   (- (:max-health @target) (:health @target)))]
      (when (pos? aid)
        (commute healer update-in [:mana] - (max 5 (/ aid 5)))
        (commute target update-in [:health] + aid)))))

(dosync (alter bilbo assoc-in [:health] 95))
(heal gandalf bilbo)
(heal gandalf bilbo)


src/demo/core.clj
(ns demo.core
  (:import (com.sun.org.apache.bcel.internal.generic BIPUSH)))
(require 'clojure.java.io)
(defn foo
  "I don't do a whole lot."
  [x]
  (println x "Hello, World!"))
;; day2 1 ;;;;;;;;;;
(defn print-logger
  [wirter]
  #(binding [*out* wirter]
     (println %)))

(def *out*-logger (print-logger *out*))
(def writer (java.io.StringWriter.))

(defn file-logger
  [file]
  #(with-open [f (clojure.java.io/writer file :append true)]
     ((print-logger f) %)))
(def log->file (file-logger "C:/coding/clojure/temp/messages.log"))

(log->file "hello")

(defn muti-logger
  [& logger-fns]
  #(doseq [f logger-fns]
     (f %)))

(def log (muti-logger
           (print-logger *out*)
           (file-logger "C:/coding/clojure/temp/messages.log")))

(defn timed-logger
  [logger]
  #(logger
     (format "[%1$tY-%1$tm-%1$te %1$tH:%1$tM:%1$tS] %2$s" (java.util.Date.) %)))

(def log-timed (timed-logger
                 (muti-logger
                   (print-logger *out*)
                   (file-logger "C:/coding/clojure/temp/messages.log"))))

(log-timed "hello hello")

(defn make-user
  [& [user-id name]]
  {:user-id (or user-id
               (str (java.util.UUID/randomUUID)))
   :name (or name
             "default-name")})

(make-user "abc" "alan")

(map-indexed vector "Clojure")

(filter (comp (partial > 25) :age)
        [{:age 21} {:age 20} {:age 34}])

(filter (partial > 25) [21,22,34])

(group-by #(rem % 3) (range 10))

(defn reduce-by
  [key-fn f init coll]
  (reduce (fn [summaries x]
            (let [k (key-fn x)]
              (assoc summaries k (f (summaries k init) x))))
          {} coll))


(def orders
  [{:product "Clock", :customer "Wile Coyote", :qty 6, :total 300}
   {:product "Dynamite", :customer "Wile Coyote", :qty 20, :total 5000}
   {:product "Shotgun", :customer "Elmer Fudd", :qty 2, :total 800}
   {:product "Shells", :customer "Elmer Fudd", :qty 4, :total 100}
   {:product "Hole", :customer "Wile Coyote", :qty 1, :total 1000}
   {:product "Anvil", :customer "Elmer Fudd", :qty 2, :total 300}
   {:product "Anvil", :customer "Wile Coyote", :qty 6, :total 900}])

(reduce-by :customer #(+ %1 (:total %2)) 0 orders)
(reduce-by :product #(conj %1 (:customer %2)) #{} orders)

(reduce-by (juxt :customer :product)
           #(+ %1 (:total %2)) 0 orders)

(defn reduce-by-in
  [keys-fn f init coll]
  (reduce (fn [summaries x]
            (let [ks (keys-fn x)]
              (assoc-in summaries ks
                        (f (get-in summaries ks init) x))))
          {} coll))
(reduce-by-in (juxt :customer :product) #(+ %1 (:total %2)) 0 orders)

(reduce #(apply assoc-in %1 %2) {}
        (reduce-by (juxt :customer :product) #(+ %1 (:total %2)) 0 orders))
;; test
( (juxt :product :customer) (first orders))
(assoc-in {} ["abc" "ddd" "eee"] 300)
;; end test

(def x (transient []))
(count x)
(def y (conj! x 1))
(count y)
(persistent! y)

(defmacro futures
  [n & exprs]
  (vec (for [_ (range n)
             expr exprs]
         `(future ~expr))))
(defmacro wait-futures
  [& args]
  `(doseq [f# (futures ~@args)]
     @f#))

(def xs (atom #{1 2 3}))
(wait-futures 1 (swap! xs (fn [v]
                            (Thread/sleep 250)
                            (println "trying 4")
                            (conj v 4)))
              (swap! xs (fn [v]
                          (Thread/sleep 500)
                          (println "trying 5")
                          (conj v 5))))

(defn get-document
  [id]
  ; ... do some work to retrieve the identified document's metadata ...
  {:url "http://www.mozilla.org/about/manifesto.en.html"
   :title "The Mozilla Manifesto"
   :mime "text/html"
   :content (future (slurp "http://www.mozilla.org/about/manifesto.en.html"))})
(def d (get-document "some-id"))
(realized? (:content d))
@(:content d)

(def p (promise))
(def a (promise))
(def b (promise))
(def c (promise))

(future
  (deliver c (+ @a @b))
  (println "Delivery complete!"))

(deliver a 15)
(deliver b 15)

@c

(def a (promise))
(def b (promise))

(future (deliver a @b))
(future (deliver b @a))

(deliver a 42)

@b

(defn call-service
  [arg1 arg2 callback-fn]
  ; ...perform service call, eventually invoking callback-fn with results...
  (future (callback-fn (+ arg1 arg2) (- arg1 arg2))))

(defn sync-fn
  [async-fn]
  (fn [& args]
    (let [result (promise)]
      (apply async-fn (conj (vec args) #(deliver result %&)))
      @result)))

((sync-fn call-service) 8 7)

(defn phone-numbers
  [string]
  (re-seq #"(\d{3})[\.-]?(\d{3})[\.-]?(\d{4})" string))

(phone-numbers "231-333-3333")

(def files (repeat 100
                   (apply str
                          (concat (repeat 1000000 \space)
                                  "Sunil: 617.555.2937, Betty: 508.555.2218"))))
(time (dorun (map phone-numbers files)))

(time (dorun (pmap phone-numbers files)))

(def files (repeat 100000
                   (apply str
                          (concat (repeat 1000 \space)
                                  "Sunil: 617.555.2937, Betty: 508.555.2218"))))
(time (->> files
           (partition-all 250)
           (pmap (fn [chunk] (doall (map phone-numbers chunk))))
           (apply concat)
           dorun))

;;watch
(defn echo-watch
  [key identity old new]
  (println key identity old "==>" new))

(def sarah (atom {:name "Sarah" :age 25}))

(add-watch sarah :echo echo-watch)

(swap! sarah update-in [:age] inc)

(def n (atom 1 :validator pos?))

(swap! n + 400)

(swap! n - 1000)

(def m {:a 1, :b 2, :c 3})

;; ref ;;;;;;;;;;;;
(defn character
  [name & {:as opts}]
  (ref (merge {:name name :items #{} :health 500}
              opts)))

(def smaug (character "Smaug" :strength 400 :items (set (range 40))))
(def bilbo (character "Bilbo" :strength 100 :health 100 ))
(def gandalf (character "Gandalf" :mana 750 :health 75  ))

;; bilbo,gandalf 分smaug 的武器 :items
(defn loot
  [from to]
  (dosync
    (when-let [item (first (:items @from))]
      (alter to update-in [:items] conj item)
      (alter from update-in [:items] disj item))))

(wait-futures 1
              (while (loot smaug bilbo))
              (while (loot smaug gandalf)))
@smaug
@bilbo
@gandalf

(map (comp count :items deref) [bilbo gandalf])

(filter (:items @bilbo) (:items @gandalf))

;;;;;;;;;;;;;;;;;;;;;
(def x (ref 0))
(time (wait-futures 1
                    (dotimes [_ 1000]
                      (dosync (alter x + (apply + (range 1000)))))
                    (dotimes [_ 1000]
                      (dosync (alter x - (apply + (range 1000)))))
                    ))

(time (wait-futures 1
                    (dotimes [_ 1000]
                      (dosync (commute x + (apply + (range 1000)))))
                    (dotimes [_ 1000]
                      (dosync (commute x - (apply + (range 1000)))))
                    ))

;;;;;;;;;;;;;;;;;;;;;;